File: 56_Merge_Intervals/merge.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {} 
};


struct less_first {
    bool operator ()(Interval const &a, Interval const &b) const {
        return a.start > b.start;
    }
};

class Solution {
	public:
    	vector<Interval> merge(vector<Interval>& intervals) {
            sort(intervals.begin(), intervals.end(), less_first());
            int idx = 0;
            while(idx < intervals.size() - 1) {
                if(intervals[idx].end < intervals[idx + 1].end &&
                   intervals[idx].end >= intervals[idx + 1].start) {
                    intervals[idx + 1].start = intervals[idx].start;
                    intervals.pop_back();
                }
                else if(intervals[idx].end >= intervals[idx + 1].end) {
                    intervals[idx + 1].start = intervals[idx].start;
                    intervals[idx + 1].end = intervals[idx].end;
                    intervals.pop_back();
                }
                else idx++;
            }
            return intervals;
    	}
};


int main() {
    Interval* a = new Interval(1, 2);
    cout << a -> start << " " << a -> end << endl;
}

File: 56_Merge_Intervals/merge.py
class Interval(object):
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        res = []
        intervals.sort(key=lambda x: x.start)
        idx = 0;

        while idx < len(intervals) - 1:
            # [1,4],[2,3] --> [1, 4]
            if intervals[idx + 1].end > intervals[idx].end >= intervals[idx + 1].start:
                intervals[idx + 1].start = intervals[idx].start
                del(intervals[idx])
            #[1,2],[2,3] --> [1, 3]
            elif intervals[idx].end >= intervals[idx + 1].end:
                intervals[idx + 1].start = intervals[idx].start
                intervals[idx + 1].end = intervals[idx].end
                del(intervals[idx])
            else:
                idx += 1
        return intervals






intervals = [Interval(4, 3), Interval(1, 2)]
sol = Solution()
sol.merge(intervals)
